src/table.js

```diff
@@ -684,10 +758,115 @@ export function __table(source, operations) {
       }))
     );
   }
-  if (primitive) source = source.map((d) => d.value);
   if (source !== input) {
     if (schema) source.schema = schema;
     if (columns) source.columns = columns;
   }
   return source;
 }
+
+function coerceRow(object, types, schema) {
+  const coerced = {};
+  for (const col of schema) {
+    const type = types.get(col.name);
+    const value = object[col.name];
+    coerced[col.name] = type === "raw" ? value : coerceToType(value, type);
+  }
+  return coerced;
+}
+
+function initKey() {
+  return {
+    other: 0,
+    boolean: 0,
+    integer: 0,
+    number: 0,
+    date: 0,
+    string: 0,
+    array: 0,
+    object: 0,
+    bigint: 0,
+    buffer: 0
+  };
+}
+
+// We need to show *all* keys present in the array of Objects
+function getAllKeys(rows) {
+  const keys = new Set();
+  for (const row of rows) {
+    // avoid crash if row is null or undefined
+    if (row) {
+      // only enumerable properties
+      for (const key in row) {
+        // only own properties
+        if (Object.prototype.hasOwnProperty.call(row, key)) {
+          // unique properties, in the order they appear
+          keys.add(key);
+        }
+      }
+    }
+  }
+  return Array.from(keys);
+}
+
+export function inferSchema(source, columns = getAllKeys(source)) {
+  const schema = [];
+  const sampleSize = 100;
+  let sample = source.slice(0, sampleSize);
+  const typeCounts = {};
+  for (const d of sample) {
+    for (const col of columns) {
+      if (!typeCounts[col]) typeCounts[col] = initKey();
+      // for json and sqlite, we already have some types, but for csv and tsv, all
+      // columns are strings here.
+      const type = typeof d[col];
+      const value = type === "string" ? d[col].trim() : d[col];
+      if (type !== "string") {
+        if (Array.isArray(value)) typeCounts[col].array++;
+        else if (value instanceof Date) typeCounts[col].date++;
+        else if (value instanceof ArrayBuffer) typeCounts[col].buffer++;
+        else if (type === "number") {
+          if (Number.isInteger(+value)) typeCounts[col].integer++;
+          else typeCounts[col].number++;
+        }
+        // bigint, boolean, or object
+        else if (type in typeCounts[col]) typeCounts[col][type]++;
+        else if (value !== null && value !== undefined) typeCounts[col].other++;
+      } else {
+        if (value === "true" || value === "false") typeCounts[col].boolean++;
+        else if (value && !isNaN(value)) {
+          if (Number.isInteger(+value)) typeCounts[col].integer++;
+          else typeCounts[col].number++;
+        } else if (/^\d+n$/.test(value)) typeCounts[col].bigint++;
+        else if (
+          value &&
+          value.match(
+            /^(([-+]\d{2})?\d{4}(-\d{1,2}(-\d{1,2})?)|(\d{1,2})\/(\d{1,2})\/(\d{2,4}))?([T ]\d{2}:\d{2}(:\d{2}(\.\d{3})?)?(Z|[-+]\d{2}:\d{2})?)?$/
+          )
```

Comment:
We should use `/regex/.test(value)` here instead of `value.match(/regex/)`. It’s faster since it doesn’t need to return any information about matched groups. Also, you don’t need the `value &&` here since you’ve already checked that `typeof value === "string"`.